New parameter `--force` on Publish command

PalmEmanuel opened the pull request:
## Contributing a feature

* [x] I have opened a new issue for the proposal, or commented on an existing one, and ensured that the Bicep maintainers are good with the design of the feature being implemented
* [x] I have included "Fixes #{issue_number}" in the PR description, so GitHub can link to the issue and close it when the PR is merged
* [x] I have appropriate test coverage of my new feature

Fixes #4757


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/Azure/bicep/pull/9745)

PalmEmanuel commented:
As far as I can tell, it's working as expected with `bicep publish` now giving an error if a module with the same tag exists in the oci registry, unless using `--force`. I implemented the module-existing-check using the `ContainerRegistryClient` (not Blob) which made it a little bit intimidating to get started on testing stuff since it requires a new mocking structure for the client unless I'm mistaken.

Can you have a look and see if the design looks good otherwise? And in that case, a pointer or two would be great on where I should get started on the testing. Thanks!

PalmEmanuel commented:
@majastrz maybe you could have a look when you have time?

majastrz commented on src/Bicep.Core/Registry/ContainerRegistryClientFactory.cs:
```diff
@@ -30,13 +30,33 @@ public ContainerRegistryBlobClient CreateAuthenticatedBlobClient(RootConfigurati
         }
 
-        public ContainerRegistryBlobClient CreateAnonymouosBlobClient(RootConfiguration configuration, Uri registryUri, string repository)
+        public ContainerRegistryBlobClient CreateAnonymousBlobClient(RootConfiguration configuration, Uri registryUri, string repository)
```
>CreateAnonymousBlobClient [](http://example.com/codeflow?start=43&length=25)

Thanks for fixing the typos!

majastrz commented on src/Bicep.Core/Registry/AzureContainerRegistryManager.cs:
```diff
@@ -103,12 +105,45 @@ public async Task PushArtifactAsync(RootConfiguration configuration, OciArtifact
+                // Check if there is an artifact with the expected tag
+                var artifact = client.GetArtifact(moduleReference.Repository, moduleReference.Tag);
+                var tagInfo = artifact.GetTagProperties(moduleReference.Tag);
+                return artifact.GetAllTagProperties().Any(a => a.Name == moduleReference.Tag);
```
>GetAllTagProperties [](http://example.com/codeflow?start=32&length=19)

The challenge with this approach is that you can verify that the tag exists, but it's not guaranteed that the tag actually points to a Bicep module.

I think `CheckModuleExists()` in `OciModuleRegistry` can just call `PullArtifactAsync` and validate that it got a manifest that belongs to a Bicep module. (It shouldn't touch the local module cache, though.)

If you get an exception with 404, then the module doesn't exist. If you get a manifest that isn't a Bicep module, then the user may want to know about that as well.

This approach would also remove the need to create separate registry clients as well. The existing blob client would be sufficient.

PalmEmanuel replied:
I didn't consider that possible false positive. Great call, I'll do that!

majastrz commented on src/Bicep.Cli.IntegrationTests/PublishCommandTests.cs:
```diff
@@ -189,7 +189,7 @@ public async Task Publish_ValidFile_ShouldSucceed(DataSet dataSet, string docume
         [DataTestMethod]
         [DynamicData(nameof(GetValidDataSets), DynamicDataSourceType.Method, DynamicDataDisplayNameDeclaringType = typeof(DataSet), DynamicDataDisplayName = nameof(DataSet.GetDisplayName))]
-        public async Task Publish_ValidArmTemplteFile_ShouldSucceed(DataSet dataSet)
+        public async Task Publish_ValidArmTemplateFile_ShouldSucceed(DataSet dataSet)
```
>Publish_ValidArmTemplateFile_ShouldSucceed [](http://example.com/codeflow?start=26&length=42)

You will need to add more tests to this class to validate the flow. (I also approved CI to run for you, so you should see results soon.)

PalmEmanuel replied:
I've added some tests, let me know if I need to change something more.

majastrz reviewed (changes requested):
I added some comments. Please take a look.

PalmEmanuel commented:
Since the tests ensured functionality by publishing twice and checking that only one module was published, I decided to modify them to also account for the `--force` option. If we'd rather have that as separate tests I can break them out.

I've also adjusted to use the blob client to pull and check manifest instead, and removed the traces of the registry client. Take a look when you have time @majastrz 🙂

majastrz commented on src/Bicep.Core/Registry/OciModuleRegistry.cs:
```diff
@@ -76,6 +76,41 @@ public override bool IsModuleRestoreRequired(OciArtifactModuleReference referenc
+                // Found no module at all
+                return false;
+            }
+            catch (RequestFailedException exception) when (exception.Message.StartsWith("The requested digest does not match the digest of the received manifest."))
```
>The requested digest does not match the digest of the received manifest. [](http://example.com/codeflow?start=89&length=72)

That's not the right exception message to catch the wrong artifact type. In the case where the OCI artifact has the wrong artifactType or has the wrong layer media type, we would throw in one of these places:
* [https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L175](https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L175)
* [https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L212](https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L212)

In addition we shouldn't assert any specific exception message since we control both the catching and the throwing of the exception. My suggestion would be to add a `Code` or `Kind` enum property to the `InvalidModuleException` class with these values: `NotSpecified`, `WrongArtifactType`, and `WrongModuleLayerMediaType` and then update the throw sites to set it. (In places not relevant to these codes, can just default to `NotSpecified`.)

(Btw, Digest mismatch on the manifest would be a bug or compromise of the registry and we throw a different exception for this at [https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L165](https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L165))

PalmEmanuel replied:
The reason for matching that message was that I got that message back when trying to overwrite a artifact of a completely different type, in this case I was trying with the `hello-world:latest` container from Docker to see what would happen, and that was the result. Not sure if that's a case intended to be caught in the mismatch error you referred.

PalmEmanuel replied:
Disregarding the case with the `hello-world` image, I added a "Kind" enum according to your suggestion and made `InvalidModuleException` public in its own file so that we can catch it and check the `Kind` property for `WrongArtifactType` or `WrongModuleLayerMediaType`.

majastrz replied:
> The reason for matching that message was that I got that message back when trying to overwrite a artifact of a completely different type, in this case I was trying with the `hello-world:latest` container from Docker to see what would happen, and that was the result. It seems that the blob client itself actually verifies that the digest matches. Not sure if that's a case intended to be caught in the mismatch error you refer to.

Ahh got it. Didn't realize that the client was already doing those checks. They are the same case, so we might be doing extra work.

Fix looks good to me also.

majastrz commented:
@PalmEmanuel sorry for the delay with the review. I added a couple more comments. Can you take a look when you get a chance?

PalmEmanuel commented:
Added some changes according to your feedback. See my comment in resolved conversation when you have time @majastrz

majastrz reviewed (approved):
:shipit:

majastrz commented:
LGTM!
